cogs/wiki.py

```diff
@@ -38,42 +44,90 @@ def __init__(self, bot: Bot):
-        async with http_session.get(url=self.url, params=params) as reply:
-            response = await reply.json()
-        if 'error' in response:
-            try:
-                info = ''.join(response['error']['info'])
-                return await ctx.send(f'Sorry, that query resulted in a search error: {info}')
-            except ValueError as e:
-                log.exception(e)
-                return await ctx.send(f'Sorry, that query resulted in a search error with no'
-                                      ' error message. Exception logged.')
-        results = response['query']['search']
-        if len(results) == 0:
-            return await ctx.send('Sorry, that query didn\'t find any article titles.')
-        urls = await self.pageids_to_urls([item['pageid'] for item in results])
-        reply = ''
-        for match, url in zip(results, urls):
-            title = match['title']
-            reply += f'\n[{title}]({url})'
-        embed = Embed(colour=Colour(0xc3c9b1),
-                      description=reply)
-        await ctx.send(embed=embed)
+        async with ctx.typing():
+            await self.refresh_titles_cache()  # fetch, or refresh, self.all_pages
+            query = ' '.join(args)
+            loop = asyncio.get_running_loop()
+            # run this CPU task in an executor to avoid blocking the bot in the meantime.
+            # functools.partial workaround is required in order to pass keyword arguments through.
+            # see: https://docs.python.org/3/library/asyncio-eventloop.html#asyncio-pass-keywords
+            with concurrent.futures.ThreadPoolExecutor() as pool:
+                results = await loop.run_in_executor(
+                    pool,
+                    functools.partial(process.extractBests,
+                                      query,
+                                      self.all_titles.keys(),
+                                      score_cutoff=75,
+                                      limit=10))  # TODO: read from config
+            if len(results) == 0:
+                return await ctx.send(f'Sorry, no matches were found for that query.')
+            pageids = [self.all_titles[item[0]] for item in results]  # map titles to IDs
+            urls = await self.pageids_to_urls(pageids)
+            reply = ''
+            for title, url in zip((result[0] for result in results), urls):
+                reply += f'\n[{title}]({url})'
+
+            embed = Embed(colour=Colour(0xc3c9b1),
+                          description=reply)
+            await ctx.send(embed=embed)
 
     @command()
     async def wikisearch(self, ctx: Context, *args):
-        """Search all articles for the given text."""
+        """Search all articles for the given text.
+        Sandbox link for this query:
+        https://cavesofqud.gamepedia.com/Special:ApiSandbox#action=query&format=json&list=search&srsearch=test&srwhat=text&srprop=snippet
+        """
```

Comment:
Query link is pretty helpful, but I think itd be nice to have a summary of the API config in the description as well (like what namespaces it searches, what algorithm, etc). same questions from wiki() apply here as well.

Reply:
I've added a bunch more comments describing what's going on with the API parameters.